core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -282,6 +317,188 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+  }
+
+  /**
+   * Helper for performance tests.  Takes the explicitly blacklisted nodes and executors; verifies
+   * that the blacklists are used efficiently to ensure scheduling is not O(numPendingTasks).
+   */
+  private def blacklistPerformanceCheck(
+      nodeBlacklist: Seq[String],
+      execBlacklist: Seq[String]): Unit = {
+    // When an executor or node is blacklisted, we want to make sure that we don't try scheduling
+    // each pending task, one by one, to discover they are all blacklisted.  This is important for
+    // performance -- if we did check each task one-by-one, then responding to a resource offer
+    // (which is usually O(1)-ish) would become O(numPendingTasks), which would slow down
+    // scheduler throughput and slow down scheduling even on healthy executors.
+    // Here, we check a proxy for the runtime -- we make sure the scheduling is short-circuited
+    // at the node or executor blacklist, so we never check the per-task blacklist.  We also make
+    // sure we don't check the node & executor blacklist for the entire taskset O(numPendingTasks)
+    // times.
+
+    taskScheduler = setupSchedulerWithMockTaskSetBlacklist()
+    // we schedule 500 tasks so we can clearly distinguish anything that is O(numPendingTasks)
+    val taskSet = FakeTask.createTaskSet(numTasks = 500, stageId = 0, stageAttemptId = 0)
+    taskScheduler.submitTasks(taskSet)
+
+    val offers = IndexedSeq(
+      new WorkerOffer("executor1", "host1", 1),
+      new WorkerOffer("executor2", "host1", 1),
+      new WorkerOffer("executor3", "host2", 1)
+    )
+    // We should check the node & exec blacklists, but only O(numOffers), not O(numPendingTasks)
+    // times.  Its O(numOffers), not exactly numOffers, because we offer resources multiple
+    // times.  An upper bound on the worst case is -- we check it once for every core at each
+    // locality level.  (We could tighten the bound a bit but that should be a good enough check.)
+    val numCoresOnAllOffers = offers.map(_.cores).sum
+    val numLocalityLevels = TaskLocality.values.size
+    val maxBlacklistChecks = numCoresOnAllOffers * numLocalityLevels
+
+    // Setup the blacklist, and get back a list of the executors & nodes that have any blacklisting
+    // (even implicit blacklisting).
+
+    nodeBlacklist.foreach { node =>
+      when(stageToMockTaskSetBlacklist(0).isNodeBlacklistedForTaskSet(node)).thenReturn(true)
+    }
+    execBlacklist.foreach { exec =>
+      when(stageToMockTaskSetBlacklist(0).isExecutorBlacklistedForTaskSet(exec)).thenReturn(true)
+    }
+    val nodesForBlacklistedExecutors = offers.filter { offer =>
+      execBlacklist.contains(offer.executorId)
+    }.map(_.host).toSet.toSeq
+    val nodesToCheck = nodeBlacklist ++ nodesForBlacklistedExecutors
+    val execsForBlacklistedNodes = offers.filter { offer =>
+      nodeBlacklist.contains(offer.host)
+    }.map(_.executorId).toSeq
+    val executorsToCheck = execBlacklist ++ execsForBlacklistedNodes
```

Comment:
This test is pretty complicated, and hard to reason about as a result.  Can you simplify fit by making a resource offer for just one executor/node, which is the executor/node that's blacklisted?  Then the list of executors / nodes to check is small (just 1), and you know exactly the number of times isNodeBlacklistedForTaskSet should be called (# locality levels) because there aren't any ordering issues?

I'm proposing this because it took me a while to read/understand this test, which makes me concerned about maintainability.